### PR TITLE
Nested WAI Applications under Scotty Routes

### PR DESCRIPTION
**Pull request by sordina**
This PR adds a new `nested` handler that allows you to place an entire WAI Application under a Scotty route.

Let me know if this implementation fits!

Simple Example:

```haskell
simpleApp :: Application
simpleApp _ respond = do
    putStrLn "I've done some IO here"
    respond $ responseLBS
        status200
        [("Content-Type", "text/plain")]
        "Hello, Web!"

main :: IO ()
main = do
  scotty 3000 $ do
    get "/simple" $ nested simpleApp
```

Reference:

* Spec included under `describe nested` in `ScottySpec.hs`.
* Examples included under `examples/nested.hs`

### CODE REVIEW

**Comment by sordina** [comment]
Looks like some of the apt-get CI commands are failing?

File: examples/nested.hs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    respond $ responseLBS
+        status200
+        [("Content-Type", "text/plain")]
+        "Hello, Web!"
+
+scottApp :: IO Application
+scottApp = scottyApp $ do
+
+    get "/" $ do
+        html $ mconcat ["<h1>Scotty, bean me up!</h1>"]
```
Comment (ocramz) [inline examples/nested.hs:23]: bean

Comment (sordina) [reply examples/nested.hs:23]: 😆

Comment (sordina) [reply examples/nested.hs]: @ocramz updated to beam

**Review by ocramz — CHANGES_REQUESTED** [review]
I think introducing 'nested' is a nice addition. @sordina could you add a test for this as well?  Edit: and a Changelog entry as well please.

**Comment by sordina** [comment]
Noticed this warning when compiling:

```
Web/Scotty/Internal/Types.hs:168:5: warning: [-Wnoncanonical-monad-instances]
    Noncanonical ‘return’ definition detected
    in the instance declaration for ‘Monad (ActionT e m)’.
    ‘return’ will eventually be removed in favour of ‘pure’
    Either remove definition for ‘return’ (recommended) or define as ‘return = pure’
    See also: https://gitlab.haskell.org/ghc/ghc/-/wikis/proposal/monad-of-no-return
    |
168 |     return = ActionT . return
```

**Comment by ocramz** [comment]
yes we have a ticket and a PR for this already!

**Comment by sordina** [comment]
@ocramz I've added a spec that matches the simple example from `examples/nested.hs`.

**Comment by ocramz** [comment]
wyt @fumieval is this good to merge?

File: Web/Scotty/Action.hs [hunk 2]
```diff
@@ -358,3 +367,14 @@ stream = ActionT . MS.modify . setContent . ContentStream
...
+
+-- | Nest a whole WAI application inside a Scotty handler.
+-- See Web.Scotty for further documentation
+nested :: (ScottyError e, MonadIO m) => Network.Wai.Application -> ActionT e m ()
+nested app = do
+  -- Is MVar really the best choice here? Not sure.
+  r <- request
+  ref <- liftIO $ newEmptyMVar
+  _ <- liftAndCatchIO $ app r (\res -> putMVar ref res >> return ResponseReceived)
+  res <- liftAndCatchIO $ readMVar ref
```
Comment (sordina) [inline Web/Scotty/Action.hs:379]: Should this catch?

Comment (ocramz) [reply Web/Scotty/Action.hs:379]: it seems 'app' could put a landmine inside the MVar so.. I think so?

**Review by ocramz — COMMENTED** [review]
It's a good PR I think

**Review by ocramz — APPROVED** [review]

**Comment by ocramz** [comment]
@sordina once you're able to add an entry to the Changelog I'll gladly merge. Thank you!

**Comment by sordina** [comment]
@ocramz Added a changelog entry under the `next` heading

**Review by ocramz — APPROVED** [review]
👍

**Comment by ocramz** [comment]
@sordina Thank you! all good 👍

**Pull request merged**